
<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Introduction to `pandas` - lab Exercise

Consider the Superstore dataset, which consist of several tables stored as .csv files.

- Products Table (products.csv)
- Orders Table (orders.csv)
- Returns Table (returns.csv)


The Products Table contains the following attributes:

- **product_id** - 	The unique identifier value for the product in the order.
- **category** - 	The product's highest level category.
- **sub_category** - 	The classification of the product within the high-level category.
- **product_name** - 	The name of the product.
- **product_cost_to_consumer** - 	The cost of each product on a single quantity basis.

Let's explore the Products table using what we have learned about Pandas so far.

Q1. Read in the products.csv file that is in the `datasets` directory into a DataFrame called `products` (You may have to go up one folder to find it, type "../" and press the TAB key)



In [1]:
# import the required libraries first!
import pandas as pd
import numpy as np

In [2]:
#Q1 Answer here
superstore_products= pd.read_csv("../lesson-1-Pandas-DataFrames/datasets/products.csv")


In [3]:
superstore_products

product_id    category sub_category  \
0     FUR-BO-10000008   Furniture    Bookcases   
1     FUR-BO-10000021   Furniture    Bookcases   
2     FUR-BO-10000035   Furniture    Bookcases   
3     FUR-BO-10000071   Furniture    Bookcases   
4     FUR-BO-10000087   Furniture    Bookcases   
...               ...         ...          ...   
2715  TEC-PH-10004760  Technology       Phones   
2716  TEC-PH-10004800  Technology       Phones   
2717  TEC-PH-10004933  Technology       Phones   
2718  TEC-PH-10004940  Technology       Phones   
2719  TEC-PH-10004959  Technology       Phones   

                                  product_name  product_cost_to_consumer  
0       Sauder Library with Doors, Traditional                    360.51  
1                 Dania Corner Shelving, Metal                     70.50  
2                 Dania Classic Bookcase, Pine                    239.04  
3                 Safco 3-Shelf Cabinet, Metal                     89.22  
4               Dania Classic Bookcase, Mobile                    392.64  
...                                        ...                       ...  
2715  Samsung Office Telephone, with Caller ID                     61.98  
2716                  Apple Headset, Full Size                     45.48  
2717                     Samsung Headset, VoIP                     42.51  
2718      Nokia Signal Booster, with Caller ID                    121.50  
2719       Motorola Audio Dock, with Caller ID                     91.89  

[2720 rows x 5 columns]

Q2. Print the first 4 rows of the `products` DataFrame.


In [4]:
#Q2 answer here
superstore_products.head(4)

product_id   category sub_category  \
0  FUR-BO-10000008  Furniture    Bookcases   
1  FUR-BO-10000021  Furniture    Bookcases   
2  FUR-BO-10000035  Furniture    Bookcases   
3  FUR-BO-10000071  Furniture    Bookcases   

                             product_name  product_cost_to_consumer  
0  Sauder Library with Doors, Traditional                    360.51  
1            Dania Corner Shelving, Metal                     70.50  
2            Dania Classic Bookcase, Pine                    239.04  
3            Safco 3-Shelf Cabinet, Metal                     89.22

Q3. Identify the shape of the `products` DataFrame (how many rows and columns?)


rows=2720
column=5

In [5]:
#Q3 answer here
superstore_products.shape

(2720, 5)

Q4. Get information about the number of null values, data types, memory usage of the DataFrame.

In [6]:
#Q4 Answer here
superstore_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2720 entries, 0 to 2719
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   product_id                2720 non-null   object 
 1   category                  2720 non-null   object 
 2   sub_category              2720 non-null   object 
 3   product_name              2720 non-null   object 
 4   product_cost_to_consumer  2720 non-null   float64
dtypes: float64(1), object(4)
memory usage: 106.4+ KB


Q5. How many products of each category are there?

In [7]:
superstore_products

product_id    category sub_category  \
0     FUR-BO-10000008   Furniture    Bookcases   
1     FUR-BO-10000021   Furniture    Bookcases   
2     FUR-BO-10000035   Furniture    Bookcases   
3     FUR-BO-10000071   Furniture    Bookcases   
4     FUR-BO-10000087   Furniture    Bookcases   
...               ...         ...          ...   
2715  TEC-PH-10004760  Technology       Phones   
2716  TEC-PH-10004800  Technology       Phones   
2717  TEC-PH-10004933  Technology       Phones   
2718  TEC-PH-10004940  Technology       Phones   
2719  TEC-PH-10004959  Technology       Phones   

                                  product_name  product_cost_to_consumer  
0       Sauder Library with Doors, Traditional                    360.51  
1                 Dania Corner Shelving, Metal                     70.50  
2                 Dania Classic Bookcase, Pine                    239.04  
3                 Safco 3-Shelf Cabinet, Metal                     89.22  
4               Dania Classic Bookcase, Mobile                    392.64  
...                                        ...                       ...  
2715  Samsung Office Telephone, with Caller ID                     61.98  
2716                  Apple Headset, Full Size                     45.48  
2717                     Samsung Headset, VoIP                     42.51  
2718      Nokia Signal Booster, with Caller ID                    121.50  
2719       Motorola Audio Dock, with Caller ID                     91.89  

[2720 rows x 5 columns]

In [8]:
#Q5 Answer here
superstore_products["category"].unique()

array(['Furniture', 'Office Supplies', 'Technology'], dtype=object)

Q6. What is the difference between using `unique()` and `nunique()`? Which one can be applied on the DataFrame and which one can be used on a Series, for example the `sub_category` column?

In [9]:
# Q6 Test out here
superstore_products["sub_category"].unique()

array(['Bookcases', 'Chairs', 'Furnishings', 'Tables', 'Appliances',
       'Art', 'Binders', 'Envelopes', 'Fasteners', 'Labels', 'Paper',
       'Storage', 'Supplies', 'Accessories', 'Copiers', 'Machines',
       'Phones'], dtype=object)

In [10]:
superstore_products.sub_category.nunique()

17

Q7. Get some descriptive statistics about the numeric variable(s): count, mean, std dev, etc...

In [11]:
#Q7 Answer here
superstore_products.describe()

product_cost_to_consumer
count               2720.000000
mean                  86.190607
std                  116.631541
min                    2.040000
25%                   14.850000
50%                   35.910000
75%                  109.530000
max                  854.040000

Q8. Using `.loc()`, find all the products in rows with index 10 - 20, showing the product_ID and product name only.

In [12]:
superstore_products

product_id    category sub_category  \
0     FUR-BO-10000008   Furniture    Bookcases   
1     FUR-BO-10000021   Furniture    Bookcases   
2     FUR-BO-10000035   Furniture    Bookcases   
3     FUR-BO-10000071   Furniture    Bookcases   
4     FUR-BO-10000087   Furniture    Bookcases   
...               ...         ...          ...   
2715  TEC-PH-10004760  Technology       Phones   
2716  TEC-PH-10004800  Technology       Phones   
2717  TEC-PH-10004933  Technology       Phones   
2718  TEC-PH-10004940  Technology       Phones   
2719  TEC-PH-10004959  Technology       Phones   

                                  product_name  product_cost_to_consumer  
0       Sauder Library with Doors, Traditional                    360.51  
1                 Dania Corner Shelving, Metal                     70.50  
2                 Dania Classic Bookcase, Pine                    239.04  
3                 Safco 3-Shelf Cabinet, Metal                     89.22  
4               Dania Classic Bookcase, Mobile                    392.64  
...                                        ...                       ...  
2715  Samsung Office Telephone, with Caller ID                     61.98  
2716                  Apple Headset, Full Size                     45.48  
2717                     Samsung Headset, VoIP                     42.51  
2718      Nokia Signal Booster, with Caller ID                    121.50  
2719       Motorola Audio Dock, with Caller ID                     91.89  

[2720 rows x 5 columns]

In [13]:
#Q8 Answer here
superstore_products.loc[10:20,["product_id","product_name"]]

product_id                            product_name
10  FUR-BO-10000210           Ikea Library with Doors, Pine
11  FUR-BO-10000243          Safco Floating Shelf Set, Pine
12  FUR-BO-10000288     Safco Classic Bookcase, Traditional
13  FUR-BO-10000321        Dania Floating Shelf Set, Mobile
14  FUR-BO-10000402  Sauder Stackable Bookrack, Traditional
15  FUR-BO-10000425    Bush Floating Shelf Set, Traditional
16  FUR-BO-10000610         Ikea Stackable Bookrack, Mobile
17  FUR-BO-10000616             Dania Corner Shelving, Pine
18  FUR-BO-10000623         Safco Library with Doors, Metal
19  FUR-BO-10000655     Dania Classic Bookcase, Traditional
20  FUR-BO-10000666            Ikea 3-Shelf Cabinet, Mobile

Q9. Using `.iloc()`, get the products in rows with index offset from 10 to 20 and the product id and product name columns.

In [14]:
#Q9 Answer here
superstore_products.iloc[10:21,[0,3]]

product_id                            product_name
10  FUR-BO-10000210           Ikea Library with Doors, Pine
11  FUR-BO-10000243          Safco Floating Shelf Set, Pine
12  FUR-BO-10000288     Safco Classic Bookcase, Traditional
13  FUR-BO-10000321        Dania Floating Shelf Set, Mobile
14  FUR-BO-10000402  Sauder Stackable Bookrack, Traditional
15  FUR-BO-10000425    Bush Floating Shelf Set, Traditional
16  FUR-BO-10000610         Ikea Stackable Bookrack, Mobile
17  FUR-BO-10000616             Dania Corner Shelving, Pine
18  FUR-BO-10000623         Safco Library with Doors, Metal
19  FUR-BO-10000655     Dania Classic Bookcase, Traditional
20  FUR-BO-10000666            Ikea 3-Shelf Cabinet, Mobile

Q10. What is the mean product cost to consumer for all products?

In [15]:
#Q10 Answer here

mean_product=superstore_products["product_cost_to_consumer"].mean()
mean_product


86.19060661764708

Q11. Find all the products in the category 'Furniture'.

In [16]:
#Q11 Answer here
superstore_products[superstore_products["category"]=="Furniture"]

product_id   category sub_category  \
0    FUR-BO-10000008  Furniture    Bookcases   
1    FUR-BO-10000021  Furniture    Bookcases   
2    FUR-BO-10000035  Furniture    Bookcases   
3    FUR-BO-10000071  Furniture    Bookcases   
4    FUR-BO-10000087  Furniture    Bookcases   
..               ...        ...          ...   
594  FUR-TA-10004745  Furniture       Tables   
595  FUR-TA-10004757  Furniture       Tables   
596  FUR-TA-10004826  Furniture       Tables   
597  FUR-TA-10004892  Furniture       Tables   
598  FUR-TA-10004957  Furniture       Tables   

                                     product_name  product_cost_to_consumer  
0          Sauder Library with Doors, Traditional                    360.51  
1                    Dania Corner Shelving, Metal                     70.50  
2                    Dania Classic Bookcase, Pine                    239.04  
3                    Safco 3-Shelf Cabinet, Metal                     89.22  
4                  Dania Classic Bookcase, Mobile                    392.64  
..                                            ...                       ...  
594      Chromcraft Conference Table, Rectangular                    731.01  
595  Barricks Training Table, with Bottom Storage                    235.56  
596             Bevis Training Table, Rectangular                    226.14  
597    Chromcraft Wood Table, with Bottom Storage                    294.36  
598             Bevis Wood Table, Fully Assembled                    288.96  

[599 rows x 5 columns]

Q12. What is the mean product cost to consumer for products in the category 'Furniture'?

In [17]:
#Q12 Answer here
superstore_products[superstore_products["category"]=="Furniture"]["product_cost_to_consumer"].mean()

157.6964774624374

Q13. Find all the products in the category "Office Supplies" that have a product cost to consumer more than 100.

In [18]:
#Q13 Answer here
superstore_products[(superstore_products["category"]=="Office Supplies") & (superstore_products["product_cost_to_consumer"] > 100)]

product_id         category sub_category  \
599   OFF-AP-10000004  Office Supplies   Appliances   
603   OFF-AP-10000108  Office Supplies   Appliances   
604   OFF-AP-10000136  Office Supplies   Appliances   
605   OFF-AP-10000304  Office Supplies   Appliances   
606   OFF-AP-10000335  Office Supplies   Appliances   
...               ...              ...          ...   
1907  OFF-ST-10004768  Office Supplies      Storage   
1909  OFF-ST-10004802  Office Supplies      Storage   
1910  OFF-ST-10004830  Office Supplies      Storage   
1911  OFF-ST-10004837  Office Supplies      Storage   
1917  OFF-ST-10004947  Office Supplies      Storage   

                            product_name  product_cost_to_consumer  
599              Cuisinart Stove, Silver                    355.77  
603         Breville Refrigerator, Black                    295.71  
604   Hamilton Beach Refrigerator, White                    328.86  
605              Hoover Microwave, White                    165.69  
606             Cuisinart Microwave, Red                    219.99  
...                                  ...                       ...  
1907           Eldon Lockers, Industrial                    115.08  
1909        Rogers File Cart, Industrial                    107.64  
1910                 Eldon Lockers, Blue                    193.98  
1911       Eldon File Cart, Single Width                    120.27  
1917      Fellowes File Cart, Industrial                    126.78  

[116 rows x 5 columns]

Q14. Sort all the products by sub category, then product cost to consumer, showing the most expensive products in each sub category first.

In [19]:
# Q14 Answer here
superstore_products.sort_values(["sub_category","product_cost_to_consumer"],ascending=[True,False])

product_id    category sub_category  \
2126  TEC-AC-10001348  Technology  Accessories   
2165  TEC-AC-10002340  Technology  Accessories   
2113  TEC-AC-10000861  Technology  Accessories   
2115  TEC-AC-10000866  Technology  Accessories   
2193  TEC-AC-10003262  Technology  Accessories   
...               ...         ...          ...   
524   FUR-TA-10001276   Furniture       Tables   
578   FUR-TA-10003637   Furniture       Tables   
525   FUR-TA-10001327   Furniture       Tables   
533   FUR-TA-10001558   Furniture       Tables   
584   FUR-TA-10004050   Furniture       Tables   

                                 product_name  product_cost_to_consumer  
2126               Logitech Router, Bluetooth                    244.86  
2165               Logitech Router, Bluetooth                    244.86  
2113                SanDisk Router, Erganomic                    213.51  
2115                 Belkin Router, Erganomic                    212.97  
2193                 Belkin Router, Erganomic                    212.97  
...                                       ...                       ...  
524   Lesro Coffee Table, with Bottom Storage                    168.84  
578   Lesro Coffee Table, with Bottom Storage                    168.84  
525      Barricks Training Table, Rectangular                    157.95  
533      Barricks Training Table, Rectangular                    157.95  
584       Lesro Coffee Table, Fully Assembled                    126.75  

[2720 rows x 5 columns]

Q15. What are the top 10 products with the highest product cost to consumer containing the word "Recycled" or "Green" in the product name?
Return your answer as a DataFrame object.

In [20]:
#Q15 Answer here
filtered_dataframe=superstore_products[superstore_products["product_name"].str.contains('Recycled|Green', case=True, regex=True)]

filtered_dataframe.sort_values("product_cost_to_consumer",ascending=False)


product_id         category sub_category  \
1615  OFF-PA-10000976  Office Supplies        Paper   
1673  OFF-PA-10003273  Office Supplies        Paper   
953   OFF-BI-10001645  Office Supplies      Binders   
1643  OFF-PA-10002047  Office Supplies        Paper   
1620  OFF-PA-10001258  Office Supplies        Paper   
...               ...              ...          ...   
1028  OFF-BI-10003475  Office Supplies      Binders   
905   OFF-BI-10000287  Office Supplies      Binders   
920   OFF-BI-10000698  Office Supplies      Binders   
923   OFF-BI-10000724  Office Supplies      Binders   
1001  OFF-BI-10002885  Office Supplies      Binders   

                                    product_name  product_cost_to_consumer  
1615        Green Bar Cards & Envelopes, Premium                     47.91  
1673        Green Bar Cards & Envelopes, Premium                     47.91  
953              Ibico Binding Machine, Recycled                     46.62  
1643       Green Bar Cards & Envelopes, 8.5 x 11                     46.23  
1620     Green Bar Cards & Envelopes, Multicolor                     45.33  
...                                          ...                       ...  
1028            Wilson Jones Index Tab, Recycled                      3.69  
905             Wilson Jones Index Tab, Recycled                      3.69  
920   Wilson Jones Hole Reinforcements, Recycled                      3.18  
923          Avery Hole Reinforcements, Recycled                      2.55  
1001         Avery Hole Reinforcements, Recycled                      2.55  

[148 rows x 5 columns]